mantidproject/mantid
# Enhance conversions between dSpacing and TOF in ConvertUnits

**Description of work.**

This change allows unit conversions between d spacing and TOF (in both directions) to be performed using the ConvertUnits algorithm as part of the powder diffraction  data reduction workflow at ISIS. These conversions use a quadratic relationship between d spacing and TOF described using three "diffractometer constants" which are typically called DIFA, DIFC and TZERO. The conversion was previously implemented in a special algorithm called AlignDetectors that only supported conversion from TOF to dSpacing.

These changes will reduce the potential for confusion around how the detector calibration is handled in Mantid and they also support a couple of processes that weren't possible previously:

- conversion of a focussed dataset from d spacing to TOF based on user selected DIFA, DIFC, TZERO values. Previously this had to be done using a DIFC based on the nominal detector positions or you had to overwrite the instrument definition with a "per bank" pseudo detector at a specific location
- it's possible to load a TOF dataset containing a calibration from GSAS into Mantid and then convert it to dSpacing and get the correct result

More specifically the changes cover:

- Changes to the Unit class and ConvertUnits algorithm to allow diffractometer constants stored in the instrument map to be used in conversions between d spacing and TOF and v.v.

- The MomentumTransfer and QSquared units have also been updated to use the difc diffractometer constant if present

- Rather than add 3 more parameters to lots of functions I have changed some of the Unit methods to take a std::unordered_map containing the detector specific parameters. I have included the 3 new diffractometer constants, l2, twoTheta, efixed and delta in this new map. The python API for converting units uses the UnitConversion class and I have left that the same though to avoid knock on effects to any python scripts out there

- New diffractometerConstants method added to SpectrumInfo and DetectorInfo to allow retrieval for a particular spectrum including averaging across detectors. Also manages situations where some of the three constants are missing.

- New algorithm created called ApplyDiffCal to read the diffractometer constants from a diffraction calibration table workspace
and write them into the instrument parameter map

- ISIS Powder diffraction scripts updated to use ApplyDiffCal and ConvertUnits instead of AlignDetectors during create_vandadium and focus actions. The separate WISH powder diffraction scripts also use AlignDetectors but am leaving those alone for now - not sure if they're used

- Added diffractometer constants to the Show Detectors screen for elastic workspaces

- Remove the Diffraction.cpp unit and replace with calls to dSpacing unit class

Although it should produce the same outputs as before where no calibrations are present, there are one or two system test ref files with updates:

- the change from AlignDetectors to ConvertUnits in the ISIS powder scripts means that spectra that failed a unit conversion get masked rather than simply having their Y data zeroed.
- the SNSPowderRedux test seems sensitive to tiny changes in the input data so some v small (~15 dp) changes in the inputs resulted in much larger changes in the output of two of the tests

**Report to:** Helen Playford

**To test:**

Run the following script. Open the "Show Detectors" screen on the right click menu of the test_ws_tof workspace and observe the calibration (DIFA, DIFC, TZERO parameters) applied on the first spectrum:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

test_ws_tof = CreateSampleWorkspace(NumBanks=1, BankPixelWidth=2, XMin=17, XMax=18, BinWidth=1)

calibration_ws = CreateEmptyTableWorkspace()
calibration_ws.addColumn("int", "detid")
calibration_ws.addColumn("double", "difc")
calibration_ws.addColumn("double", "difa")
calibration_ws.addColumn("double", "tzero")
calibration_ws.addRow([4, 2.0, 3.0, 1.0])

ApplyDiffCal(InstrumentWorkspace=test_ws_tof, CalibrationWorkspace=calibration_ws)

# TOF = difa*d*d  + difc*d + tzero, TOF=17 gives d spacing=2
test_ws_d = ConvertUnits(InputWorkspace=test_ws_tof, Target="dSpacing")

dataX = test_ws_d.dataX(0)
assert(dataX[0]==2.0)

# convert back using calibration that is still attached to workspace
test_ws_tof2 = ConvertUnits(InputWorkspace=test_ws_d, Target="TOF")

(result, messages) = CompareWorkspaces(test_ws_tof,test_ws_tof2,1E-7)

print("Result: {}".format(result))
```

Fixes #28938 .

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
